File: tools/archive-articles.py
#!/usr/bin/env python3
"""
ARTICLE ARCHIVER

Retrieves articles from Sanity and exports them as:
- [slug].md: Article body converted from Portable Text to Markdown
- [slug].json: All metadata (excluding sources field)

Usage:
    python tools/archive-articles.py [slug] [--output-dir archive/Articles]

Arguments:
    slug            Optional: archive only this article slug. If omitted, archives all.

Options:
    --output-dir    Output directory for archived articles (default: archive/Articles)
"""

import json
import sys
import subprocess
import os
from pathlib import Path
from datetime import datetime


def portable_text_to_markdown(blocks):
    """
    Convert portable text blocks to markdown inline.
    This is a simplified version to avoid subprocess overhead.
    """
    if not blocks:
        return ""

    def resolve_mark_defs(children, mark_defs):
        mark_lookup = {}
        if mark_defs:
            for mark_def in mark_defs:
                mark_key = mark_def.get("_key")
                if mark_key:
                    mark_lookup[mark_key] = mark_def
        return mark_lookup

    def render_inline_children(children, mark_defs):
        mark_lookup = resolve_mark_defs(children, mark_defs)
        result = []

        for child in children:
            if child.get("_type") != "span":
                continue

            text = child.get("text", "")
            marks = child.get("marks", [])

            link_url = None
            non_link_marks = []
            for mark in marks:
                if mark in mark_lookup:
                    mark_def = mark_lookup[mark]
                    if mark_def.get("_type") == "link":
                        link_url = mark_def.get("href", "")
                else:
                    non_link_marks.append(mark)

            formatted_text = text
            if "code" in non_link_marks:
                formatted_text = f"`{formatted_text}`"
            if "em" in non_link_marks:
                formatted_text = f"*{formatted_text}*"
            if "strong" in non_link_marks:
                formatted_text = f"**{formatted_text}**"

            if link_url:
                formatted_text = f"[{formatted_text}]({link_url})"

            result.append(formatted_text)

        return "".join(result)

    def render_block(block):
        block_type = block.get("_type", "block")

        if block_type == "block":
            style = block.get("style", "normal")
            children = block.get("children", [])
            mark_defs = block.get("markDefs", [])

            inline_text = render_inline_children(children, mark_defs)

            if style == "normal":
                return inline_text
            elif style == "h1":
                return f"# {inline_text}"
            elif style == "h2":
                return f"## {inline_text}"
            elif style == "h3":
                return f"### {inline_text}"
            elif style == "h4":
                return f"#### {inline_text}"
            elif style == "h5":
                return f"##### {inline_text}"
            elif style == "h6":
                return f"###### {inline_text}"
            elif style == "blockquote":
                lines = inline_text.split('\n')
                return '\n'.join(f"> {line}" for line in lines)
            else:
                return inline_text

        elif block_type == "image":
            alt = block.get("alt", "")
            asset_ref = block.get("asset", {}).get("_ref", "")
            caption = block.get("caption", "")
            credit = block.get("credit", "")

            parts = [f'alt="{alt}"']
            if asset_ref:
                parts.append(f'asset="{asset_ref}"')
            if caption:
                parts.append(f'caption="{caption}"')
            if credit:
                parts.append(f'credit="{credit}"')

            return f"<!-- IMAGE {' '.join(parts)} -->"

        elif block_type == "code":
            code_content = block.get("code", "")
            language = block.get("language", "text")
            return f"```{language}\n{code_content}\n```"

        elif block_type == "mermaidDiagram":
            code_content = block.get("code", "")
            return f"```mermaid\n{code_content}\n```"

        elif block_type == "table":
            rows = block.get("rows", [])
            if not rows:
                return ""

            markdown_rows = []
            for idx, row in enumerate(rows):
                cells = row.get("cells", [])
                row_text = " | ".join(str(cell) for cell in cells)
                markdown_rows.append(f"| {row_text} |")

                if idx == 0:
                    separator = " | ".join(["---"] * len(cells))
                    markdown_rows.append(f"| {separator} |")

            return "\n".join(markdown_rows)

        return ""

    markdown_lines = []
    for block in blocks:
        rendered = render_block(block)
        if rendered:
            markdown_lines.append(rendered)

    result = []
    for i, line in enumerate(markdown_lines):
        result.append(line)
        if i < len(markdown_lines) - 1:
            next_line = markdown_lines[i + 1]
            is_current_list = line.lstrip().startswith(("- ", "1. "))
            is_next_list = next_line.lstrip().startswith(("- ", "1. "))

            if not (is_current_list and is_next_list):
                result.append("")

    final_result = []
    prev_blank = False
    for line in result:
        if line.strip() == "":
            if not prev_blank:
                final_result.append("")
                prev_blank = True
        else:
            final_result.append(line)
            prev_blank = False

    return "\n".join(final_result).strip()


def query_sanity(slug=None):
    """
    Query Sanity for articles.

    Args:
        slug: Optional article slug to fetch single article

    Returns:
        list of article documents
    """
    from sanity import Client

    api_version = '2023-12-01'
    dataset = os.getenv('SANITY_DATASET', 'production')
    project_id = os.getenv('SANITY_PROJECT_ID')
    token = os.getenv('SANITY_API_TOKEN')

    if not project_id:
        print("Error: SANITY_PROJECT_ID environment variable not set", file=sys.stderr)
        sys.exit(1)

    client = Client(
        project_id=project_id,
        dataset=dataset,
        api_version=api_version,
        token=token,
    )

    if slug:
        query = f'*[_type == "article" && slug.current == "{slug}"]'
    else:
        query = '*[_type == "article"] | order(publishedAt desc)'

    try:
        results = client.fetch(query)
        return results if isinstance(results, list) else [results]
    except Exception as e:
        print(f"Error querying Sanity: {e}", file=sys.stderr)
        sys.exit(1)


def fetch_articles_via_cli(slug=None):
    """
    Fallback: Use sanity CLI to fetch articles as JSON.
    Requires: `sanity documents list --query '*[_type == "article"]' > articles.json`
    """
    try:
        if slug:
            query = f'*[_type == "article" && slug.current == "{slug}"]'
        else:
            query = '*[_type == "article"] | order(publishedAt desc)'

        # Use sanity CLI
        result = subprocess.run(
            ['sanity', 'documents', 'list', '--query', query, '--format', 'json'],
            capture_output=True,
            text=True,
            check=False
        )

        if result.returncode != 0:
            print(f"Warning: sanity CLI failed: {result.stderr}", file=sys.stderr)
            return []

        data = json.loads(result.stdout)
        return data if isinstance(data, list) else [data]

    except (FileNotFoundError, json.JSONDecodeError) as e:
        print(f"Error using sanity CLI: {e}", file=sys.stderr)
        return []


def archive_article(article, output_dir):
    """
    Archive a single article as .md and .json files.

    Args:
        article: Article document from Sanity
        output_dir: Path to output directory

    Returns:
        (md_file, json_file) tuple of output paths
    """
    slug = article.get("slug", {}).get("current", "")
    if not slug:
        print(f"Warning: Article has no slug: {article.get('_id')}", file=sys.stderr)
        return None, None

    output_dir = Path(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)

    # Convert body to markdown
    body = article.get("body", [])
    markdown_content = portable_text_to_markdown(body)

    # Prepare metadata (exclude sources, body, _rev, _createdAt)
    metadata = {}
    exclude_fields = {"body", "sources", "_rev", "_createdAt"}

    for key, value in article.items():
        if key not in exclude_fields:
            metadata[key] = value

    # Write markdown file
    md_file = output_dir / f"{slug}.md"
    with open(md_file, 'w', encoding='utf-8') as f:
        f.write(markdown_content)
    print(f"Created: {md_file}")

    # Write metadata JSON file
    json_file = output_dir / f"{slug}.json"
    with open(json_file, 'w', encoding='utf-8') as f:
        json.dump(metadata, f, indent=2, default=str, ensure_ascii=False)
    print(f"Created: {json_file}")

    return md_file, json_file


def main():
    slug = None
    output_dir = "archive/Articles"

    # Parse arguments
    for arg in sys.argv[1:]:
        if arg.startswith("--output-dir="):
            output_dir = arg.split("=", 1)[1]
        elif arg.startswith("--output-dir"):
            # Handle `--output-dir path` format
            pass
        elif not arg.startswith("--"):
            slug = arg

    print(f"Archiving articles to: {output_dir}", file=sys.stderr)
    if slug:
        print(f"Fetching single article: {slug}", file=sys.stderr)
    else:
        print("Fetching all articles...", file=sys.stderr)

    # Fetch articles
    try:
        articles = query_sanity(slug)
    except ImportError:
        print("Note: python-sanity library not installed, trying CLI fallback...", file=sys.stderr)
        articles = fetch_articles_via_cli(slug)

    if not articles:
        print("No articles found", file=sys.stderr)
        sys.exit(1)

    print(f"Found {len(articles)} article(s)", file=sys.stderr)

    # Archive each article
    archived = []
    for article in articles:
        md_file, json_file = archive_article(article, output_dir)
        if md_file and json_file:
            archived.append({
                "slug": article.get("slug", {}).get("current", ""),
                "title": article.get("title", ""),
                "markdown": str(md_file),
                "metadata": str(json_file),
            })

    print(f"\nArchived {len(archived)} article(s)", file=sys.stderr)
    for item in archived:
        print(f"  - {item['slug']}: {item['title']}", file=sys.stderr)


if __name__ == '__main__':
    main()

File: tools/archive-articles.ts
#!/usr/bin/env node

/**
 * ARTICLE ARCHIVER
 *
 * Retrieves articles from Sanity and exports them as:
 * - [slug].md: Article body converted from Portable Text to Markdown
 * - [slug].json: All metadata (excluding sources field)
 *
 * Usage:
 *   npx ts-node tools/archive-articles.ts [slug]
 *   node tools/archive-articles.mjs [slug]
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

/**
 * Convert Portable Text blocks to Markdown
 */
function portableTextToMarkdown(blocks: any[]): string {
  if (!blocks || blocks.length === 0) return '';

  interface MarkDef {
    _key: string;
    _type: string;
    href?: string;
  }

  interface Span {
    _type: string;
    text: string;
    marks?: string[];
  }

  const resolveMarkDefs = (
    children: Span[],
    markDefs: MarkDef[] = []
  ): Record<string, MarkDef> => {
    const lookup: Record<string, MarkDef> = {};
    if (markDefs) {
      for (const markDef of markDefs) {
        if (markDef._key) {
          lookup[markDef._key] = markDef;
        }
      }
    }
    return lookup;
  };

  const renderInlineChildren = (children: Span[], markDefs: MarkDef[] = []): string => {
    const markLookup = resolveMarkDefs(children, markDefs);
    const result: string[] = [];

    for (const child of children) {
      if (child._type !== 'span') continue;

      let text = child.text || '';
      const marks = child.marks || [];

      // Separate link marks from other marks
      let linkUrl: string | null = null;
      const nonLinkMarks: string[] = [];

      for (const mark of marks) {
        if (mark in markLookup) {
          const markDef = markLookup[mark];
          if (markDef._type === 'link') {
            linkUrl = markDef.href || '';
          }
        } else {
          nonLinkMarks.push(mark);
        }
      }

      // Apply non-link marks (order: code, em, strong for proper nesting)
      let formattedText = text;
      if (nonLinkMarks.includes('code')) {
        formattedText = `\`${formattedText}\``;
      }
      if (nonLinkMarks.includes('em')) {
        formattedText = `*${formattedText}*`;
      }
      if (nonLinkMarks.includes('strong')) {
        formattedText = `**${formattedText}**`;
      }

      // Apply link
      if (linkUrl) {
        formattedText = `[${formattedText}](${linkUrl})`;
      }

      result.push(formattedText);
    }

    return result.join('');
  };

  const renderBlock = (block: any): string => {
    const blockType = block._type || 'block';

    if (blockType === 'block') {
      const style = block.style || 'normal';
      const children = block.children || [];
      const markDefs = block.markDefs || [];

      const inlineText = renderInlineChildren(children, markDefs);

      switch (style) {
        case 'h1':
          return `# ${inlineText}`;
        case 'h2':
          return `## ${inlineText}`;
        case 'h3':
          return `### ${inlineText}`;
        case 'h4':
          return `#### ${inlineText}`;
        case 'h5':
          return `##### ${inlineText}`;
        case 'h6':
          return `###### ${inlineText}`;
        case 'blockquote': {
          const lines = inlineText.split('\n');
          return lines.map((line) => `> ${line}`).join('\n');
        }
        default:
          return inlineText;
      }
    }

    if (blockType === 'image') {
      const alt = block.alt || '';
      const assetRef = block.asset?._ref || '';
      const caption = block.caption || '';
      const credit = block.credit || '';

      const parts = [`alt="${alt}"`];
      if (assetRef) parts.push(`asset="${assetRef}"`);
      if (caption) parts.push(`caption="${caption}"`);
      if (credit) parts.push(`credit="${credit}"`);

      return `<!-- IMAGE ${parts.join(' ')} -->`;
    }

    if (blockType === 'code') {
      const codeContent = block.code || '';
      const language = block.language || 'text';
      return `\`\`\`${language}\n${codeContent}\n\`\`\``;
    }

    if (blockType === 'mermaidDiagram') {
      const codeContent = block.code || '';
      return `\`\`\`mermaid\n${codeContent}\n\`\`\``;
    }

    if (blockType === 'table') {
      const rows = block.rows || [];
      if (!rows.length) return '';

      const markdownRows: string[] = [];
      for (let idx = 0; idx < rows.length; idx++) {
        const row = rows[idx];
        const cells = row.cells || [];
        const rowText = cells.join(' | ');
        markdownRows.push(`| ${rowText} |`);

        if (idx === 0) {
          const separator = cells.map(() => '---').join(' | ');
          markdownRows.push(`| ${separator} |`);
        }
      }

      return markdownRows.join('\n');
    }

    return '';
  };

  const markdownLines = blocks.map(renderBlock).filter(Boolean);

  // Add spacing between blocks
  const result: string[] = [];
  for (let i = 0; i < markdownLines.length; i++) {
    result.push(markdownLines[i]);
    if (i < markdownLines.length - 1) {
      const nextLine = markdownLines[i + 1];
      const isCurrentList = /^[-*\d+.]/.test(markdownLines[i].trimStart());
      const isNextList = /^[-*\d+.]/.test(nextLine.trimStart());

      if (!(isCurrentList && isNextList)) {
        result.push('');
      }
    }
  }

  // Remove excessive blank lines
  const finalResult: string[] = [];
  let prevBlank = false;
  for (const line of result) {
    if (line.trim() === '') {
      if (!prevBlank) {
        finalResult.push('');
        prevBlank = true;
      }
    } else {
      finalResult.push(line);
      prevBlank = false;
    }
  }

  return finalResult.join('\n').trim();
}

/**
 * Archive a single article
 */
async function archiveArticle(
  article: any,
  outputDir: string
): Promise<{ md: string; json: string } | null> {
  const slug = article.slug?.current;
  if (!slug) {
    console.error(`Warning: Article has no slug: ${article._id}`);
    return null;
  }

  // Create output directory
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }

  // Convert body to markdown
  const body = article.body || [];
  const markdownContent = portableTextToMarkdown(body);

  // Prepare metadata (exclude sources, body, _rev, _createdAt)
  const metadata: any = {};
  const excludeFields = new Set(['body', 'sources', '_rev', '_createdAt']);

  for (const [key, value] of Object.entries(article)) {
    if (!excludeFields.has(key)) {
      metadata[key] = value;
    }
  }

  // Write markdown file
  const mdFile = path.join(outputDir, `${slug}.md`);
  fs.writeFileSync(mdFile, markdownContent, 'utf-8');
  console.log(`Created: ${mdFile}`);

  // Write metadata JSON file
  const jsonFile = path.join(outputDir, `${slug}.json`);
  fs.writeFileSync(jsonFile, JSON.stringify(metadata, null, 2), 'utf-8');
  console.log(`Created: ${jsonFile}`);

  return { md: mdFile, json: jsonFile };
}

/**
 * Main function
 */
async function main() {
  let slug: string | null = null;
  let outputDir = 'archive/Articles';

  // Parse arguments
  for (let i = 2; i < process.argv.length; i++) {
    const arg = process.argv[i];
    if (arg.startsWith('--output-dir=')) {
      outputDir = arg.split('=')[1];
    } else if (!arg.startsWith('--')) {
      slug = arg;
    }
  }

  console.error(`Archiving articles to: ${outputDir}`);
  if (slug) {
    console.error(`Fetching single article: ${slug}`);
  } else {
    console.error('Fetching all articles...');
  }

  try {
    // Import Sanity client from project
    const { default: sanityClient } = await import('../src/lib/sanity/lib/client.ts');

    // Build GROQ query
    let query = '*[_type == "article"';
    if (slug) {
      query += ` && slug.current == "${slug}"`;
    }
    query +=
      '] | order(publishedAt desc) {_id, title, slug, description, leadParagraph, body, mainImage, hasEmbeddedVideo, videoLink, imageGallery, author, publishedAt, updatedAt, eventDate, location, categories, tags, keywords, featured, breakingNews, isFieldReport, needsReview, allowComments, relatedArticles, linkedPitch, deletionRequest, methodology, correction, reviewedBy, seo, faqs, readingTimeMinutes}';

    const articles = await sanityClient.fetch(query);
    const articleList = Array.isArray(articles) ? articles : [articles];

    if (articleList.length === 0) {
      console.error('No articles found');
      process.exit(1);
    }

    console.error(`Found ${articleList.length} article(s)`);

    // Archive each article
    const archived: Array<{ slug: string; title: string; markdown: string; metadata: string }> =
      [];

    for (const article of articleList) {
      const result = await archiveArticle(article, outputDir);
      if (result) {
        archived.push({
          slug: article.slug?.current || '',
          title: article.title || '',
          markdown: result.md,
          metadata: result.json,
        });
      }
    }

    console.error(`\nArchived ${archived.length} article(s)`);
    for (const item of archived) {
      console.error(`  - ${item.slug}: ${item.title}`);
    }
  } catch (error) {
    console.error('Error:', error instanceof Error ? error.message : String(error));
    process.exit(1);
  }
}

main();
